### ISSUE TITLE
Branch and Price. Unfeasible branches

### ISSUE THREAD

**Issue by pambus** [SCIP-Interfaces/PySCIPOpt#443]
Hi everyone,
First of all, thank you very much for helping us with PySCIPOpt. 

I am using SCIP to implement a branch and price algorithm (with binary variables). Currently, I have a Pricer and a Branchrule (I don't have a Conshdlr). My problem is that it seems like I am only considering one branch of the tree (and not the other branches). 

For example... 
-Model before the first branching:   
```
 0 <= t_q[0,0] <= 1
 0 <= t_q[1,0] <= 1
 0 <= t_q[2,0] <= 1
```

-Model after the first branching:  
```
 0 <= t_q[0,0] <= 1
 0 <= t_q[1,0] <= 0
 0 <= t_q[2,0] <= 1
```

-Model after the second branching:
```
 0 <= t_q[0,0] <= 1
 0 <= t_q[1,0] <= 0
 0 <= t_q[2,0] <= 0
```
And it ends here, but this is not what I want. I would like to explore all the possibilities (0 <= t_q[1,0] <= 1 and the rest fixed to 0, etc.).

I am using a very simple branching mentioned in #395. Here is my code simplified.
```
class CutPricer(Pricer):
    def pricerredcost(self):       
        #(subMIP code) 

        subMIP.setPresolve(SCIP_PARAMSETTING.OFF)

        #(code to calculate reduced costs)

        if objval >1e-08:
            currentNumVar = len(self.data['var'])
            newVar = self.model.addVar("NewPattern_" + str(currentNumVar), vtype = "C", obj = 0, pricedVar = True) 
            for i, c in enumerate(self.data['cons']):
                self.model.addConsCoeff(c, newVar, new_column[i]*self.data['coef_obj'][i]) 

        return {'result':SCIP_RESULT.SUCCESS}
    
    def pricerinit(self):
        for i, c in enumerate(self.data['cons']):
            self.data['cons'][i] = self.model.getTransformedCons(c)

class MyVarBranching(Branchrule): 
    def __init__(self, model):
        self.model = model

    def branchexeclp(self, allowaddcons): #executes branching rule for fractional LP solution

        (
            lpcands,
            lpcandssol,
            lpcandsfrac,
            nlpcands,
            npriolpcands,
            nfracimplvars,
        ) = self.model.getLPBranchCands()

        integral = lpcands[0]
        self.model.branchVar(integral) 
        return {"result": SCIP_RESULT.BRANCHED}

def mip(m,n):
    model = Model("BP")

    model.setPresolve(SCIP_PARAMSETTING.OFF)
    model.setHeuristics(SCIP_PARAMSETTING.OFF)
    model.disablePropagation()

    pricer = CutPricer() 
    model.includePricer(pricer, "StrategyPricer", "Pricer to identify new defender strategies",delay = True) 

    my_branchrule = MyVarBranching(model)

    model.includeBranchrule(
        my_branchrule,
        "test branch",
        "test branching and probing and lp functions",
        priority=10000000,
        maxdepth=-1,
        maxbounddist=1,
    )

    varNames = []
    patterns = []

    #(Code to create initial feasible columns.) 
    #(Code to write model)

    return model
```

I have looked at codes like test_pricer.py and test_coloring.py, but I still don't know how to solve my problem. 
This is a problem for me because I can't obtain the optimal solution. Instead, I get just a feasible solution.

Thank you very much.

Best,
Pamela

**Comment by fserra** [comment]
Hola Pamela

I don't really know what is going on. Could you share the full code (or better a small working example) that I can use to reproduce the behavior you describe, please?

**Comment by pambus** [comment]
I think my main problem is with the branching. Maybe I misunderstood something. For example, in test_branch_probing_lp.py, if I change the lb of x0 to -2.1, the result of x0 is -2.1 (but I believe it should be integral since getLPBranchCands is being used).
The same happens if lb of x0 is -1.9 (the result of x0 is -1.9). 
If I keep the lb of x0 as -2.0, and I erase the Branchrule, I get the same result as the one obtained with the branch rule.

I attach the code, which is the same as the test_branch_probing_lp.py, except that I changed the lb.

```
from pyscipopt import Model, Branchrule, SCIP_RESULT, quicksum


class MyBranching(Branchrule):

    def __init__(self, model, cont):
        self.model = model
        self.cont = cont
        self.count = 0
        self.was_called_val = False
        self.was_called_int = False

    def branchexeclp(self, allowaddcons):
        self.count += 1
        if self.count >= 2:
            return {"result": SCIP_RESULT.DIDNOTRUN}
        assert allowaddcons

        assert not self.model.inRepropagation()
        assert not self.model.inProbing()
        self.model.startProbing()
        assert not self.model.isObjChangedProbing()
        self.model.fixVarProbing(self.cont, 2.0)
        self.model.constructLP()
        self.model.solveProbingLP()
        self.model.getLPObjVal()
        self.model.endProbing()

        self.integral = self.model.getLPBranchCands()[0][0]

        if self.count == 1:
            down, eq, up = self.model.branchVarVal(self.cont, 1.3)
            self.model.chgVarLbNode(down, self.cont, -1.5)
            self.model.chgVarUbNode(up, self.cont, 3.0)
            self.was_called_val = True
            down2, eq2, up2 = self.model.branchVar(self.integral)
            self.was_called_int = True
            self.model.createChild(6, 7)
            return {"result": SCIP_RESULT.BRANCHED}



def test_branching():
    m = Model()
    m.setIntParam("presolving/maxrounds", 0)
    #m.setLongintParam("lp/rootiterlim", 3)
    m.setRealParam("limits/time", 60)

    x0 = m.addVar(lb=-2.1, ub=4)
    r1 = m.addVar()
    r2 = m.addVar()
    y0 = m.addVar(lb=3)
    t = m.addVar(lb=None)
    l = m.addVar(vtype="I", lb=-9, ub=18)
    u = m.addVar(vtype="I", lb=-3, ub=99)

    more_vars = []
    for i in range(1000):
        more_vars.append(m.addVar(vtype="I", lb= -12, ub=40))
        m.addCons(quicksum(v for v in more_vars) <= (40 - i) * quicksum(v for v in more_vars[::2]))

    for i in range(1000):
        more_vars.append(m.addVar(vtype="I", lb= -52, ub=10))
        m.addCons(quicksum(v for v in more_vars[50::2]) <= (40 - i) * quicksum(v for v in more_vars[405::2]))



    m.addCons(r1 >= x0)
    m.addCons(r2 >= -x0)
    m.addCons(y0 == r1 +r2)
    #m.addCons(t * l + l * u >= 4)
    m.addCons(t + l + 7* u <= 300)
    m.addCons(t >= quicksum(v for v in more_vars[::3]) - 10 * more_vars[5] + 5* more_vars[9])
    m.addCons(more_vars[3] >= l + 2)
    m.addCons(7 <= quicksum(v for v in more_vars[::4]) - x0)
    m.addCons(quicksum(v for v in more_vars[::2]) + l <= quicksum(v for v in more_vars[::4]))


    m.setObjective(t - quicksum(j*v for j, v in enumerate(more_vars[20:-40])))
    #m.addCons(t >= r1 * (r1 - x0) + r2 * (r2 + x0))

    my_branchrule = MyBranching(m, x0)
    m.includeBranchrule(my_branchrule, "test branch", "test branching and probing and lp functions",
                        priority=10000000, maxdepth=3, maxbounddist=1)

    
    m.optimize()
    print(m.getObjVal())

    print("x0", m.getVal(x0))
    print("r1", m.getVal(r1))
    print("r2", m.getVal(r2))
    print("y0", m.getVal(y0))
    print("t", m.getVal(t))

    assert my_branchrule.was_called_val
    assert my_branchrule.was_called_int

if __name__ == "__main__":
    test_branching()
```

Thank you very much

**Comment by fserra** [comment]
mmmm so `x0` is a continuous variable because its `vtype` was not specified when creating it:
```python
    x0 = m.addVar(lb=-2.1, ub=4)
```
 The documentation of [SCIPbranchVarVal](https://www.scipopt.org/doc/html/group__PublicBranchingMethods.php#ga82ced9a0be0df3fcb51b0dd514e6bdd8) says

> for continuous variables with relative domain width larger epsilon, x' must not be one of the bounds; two child nodes (x <= x', x >= x') are created;

so the lower bound in -2.1 and when branching it uses `1.3` for val
```python
            down, eq, up = self.model.branchVarVal(self.cont, 1.3)
```

so it looks fine to me. Does this explain what you observe in your application?

**Comment by pambus** [comment]
Thank you very much. Your explanation helped me understand how the branching worked. 
Regarding my problem, I realized I was adding global variables instead of local variables (and I needed local variables), so I am going to fix that and I hope that will solve my problem.

Best,

**Comment by fserra** [comment]
nice! after you check (and the problem is solved) can you close the issue, please? Gracias!

**Comment by pambus** [comment]
Yess!, of course. Thank you for your quick response.

**Comment by pambus** [comment]
Hi @fserra ,
I realized that I needed to do Farkas pricing to not have unfeasible branches (unless they were really unfeasible).
Context (in case someone has the same problem): In the pricing problem I was generating new continuous variables. After finding a positive reduced cost, I was trying to branch on binary variables. The problem was that the columns I had generated, were not enough (in some cases) to make every feasible branch, feasible.

Best,
